processing/src/main/java/io/druid/segment/CompressedPools.java

```diff
@@ -77,41 +102,245 @@ public BufferRecycler get()
+                    ByteBuffer theBuffer = compressionBufferSupplier.get();
+                    theBuffer.clear();
+                    return theBuffer;
+                  }
+                }
+              );
+          }
+        },
+        0,
+        LEMIRE_FASTPFOR_CODEC_POOL_MAX_CACHE
+    );
+  }
+
+  private static final NonBlockingPool<ByteBuffer> bigEndByteBufPool =
+      makeBufferPool("bigEndByteBufPool", BUFFER_SIZE, ByteOrder.BIG_ENDIAN);
+
+  private static final NonBlockingPool<ByteBuffer> littleBigEndByteBufPool =
+      makeBufferPool("littleBigEndByteBufPool", SMALLER_BUFFER_SIZE, ByteOrder.BIG_ENDIAN);
+
+  private static final NonBlockingPool<ByteBuffer> littlestBigEndByteBufPool =
+      makeBufferPool("littlestBigEndByteBufPool", SMALLEST_BUFFER_SIZE, ByteOrder.BIG_ENDIAN);
+
+  private static final NonBlockingPool<ByteBuffer> littleEndByteBufPool =
+      makeBufferPool("littleEndByteBufPool", BUFFER_SIZE, ByteOrder.LITTLE_ENDIAN);
+
+  private static final NonBlockingPool<ByteBuffer> littlerEndByteBufPool =
+      makeBufferPool("littlerEndByteBufPool", SMALLER_BUFFER_SIZE, ByteOrder.LITTLE_ENDIAN);
+
+  private static final NonBlockingPool<ByteBuffer> littlestEndByteBufPool =
+      makeBufferPool("littlestEndByteBufPool", SMALLEST_BUFFER_SIZE, ByteOrder.LITTLE_ENDIAN);
+
+  private static final NonBlockingPool<int[]> shapeshiftIntsDecodedValuesArrayPool =
+      makeIntArrayPool(
+          "shapeshiftIntsDecodedValuesArrayPool",
+          INT_ARRAY_SIZE,
+          INT_DECODED_ARRAY_POOL_MAX_CACHE
+      );
+
+  private static final NonBlockingPool<int[]> shapeshiftIntsEncodedValuesArrayPool =
+      makeIntArrayPool(
+          "shapeshiftIntsEncodedValuesArrayPool",
+          INT_ARRAY_SIZE + ENCODED_INTS_SHOULD_BE_ENOUGH,
+          INT_ENCODED_ARRAY_POOL_MAX_CACHE
+      );
+
+  private static final NonBlockingPool<int[]> shapeshiftSmallerIntsDecodedValuesArrayPool =
+      makeIntArrayPool(
+          "shapeshiftSmallerIntsDecodedValuesArrayPool",
+          SMALLER_INT_ARRAY_SIZE,
+          INT_DECODED_ARRAY_POOL_MAX_CACHE
+      );
+
+  private static final NonBlockingPool<int[]> shapeshiftSmallerIntsEncodedValuesArrayPool =
+      makeIntArrayPool(
+          "shapeshiftSmallerIntsEncodedValuesArrayPool",
+          SMALLER_INT_ARRAY_SIZE + ENCODED_INTS_SHOULD_BE_ENOUGH,
+          INT_ENCODED_ARRAY_POOL_MAX_CACHE
+      );
+
+  private static final NonBlockingPool<int[]> shapeshiftSmallestIntsDecodedValuesArrayPool =
+      makeIntArrayPool(
+          "shapeshiftSmallestIntsDecodedValuesArrayPool",
+          SMALLEST_INT_ARRAY_SIZE,
+          INT_DECODED_ARRAY_POOL_MAX_CACHE
+      );
+
+  private static final NonBlockingPool<int[]> shapeshiftSmallestIntsEncodedValuesArrayPool =
+      makeIntArrayPool(
+          "shapeshiftSmallestIntsEncodedValuesArrayPool",
+          SMALLEST_INT_ARRAY_SIZE + ENCODED_INTS_SHOULD_BE_ENOUGH,
+          INT_ENCODED_ARRAY_POOL_MAX_CACHE
+      );
```

Comment:
Is it worth it creating all the object pools for int[] which is always on-heap and fast to allocate when they are not long lived as appear to be the case? Did you find performance degradation without these pools?

Its a different story when they are Direct ByteBuffer whose allocation/deallocation is usually expensive.

Reply:
I think this is probably worth revisiting, I pooled these at the same time as I was trying to be more disciplined about the lifetime of objects I needed, and query metrics on our test cluster were significantly better, but it also could have been other changes.

Not pooling definitely puts more work on the garbage collector; let's consider a full scan of a column with 3M rows using the largest block size, which maps to 184 chunk and where all chunks are encoded with FastPFOR. A 64k value array is needed for the lifetime of the column to hold the currently decoded chunk. Additionally, during each chunk transition we need another 65k temporary value array to copy the encoded values of the chunk to, and a FastPFOR codec object which weighs in around 1M on heap (plus 
~264k direct buffers) to decode the 65k array contents into the 64k array. It is too heavy to just allocate one of each of these up front and keep these alive for the lifetime of the column, and the column only needs them for a very short period, so everything besides the 64k value array needs to be released somehow or it makes things like merging and compaction very hard to do because of huge heap requirements. Additionally, the FastPFOR codec objects definitely need to be pooled, if only because of the direct buffers each instantiation also allocates internally. 

If we just consider the value array and temp arrays, that is 185 arrays allocated and released (each requiring a contiguous chunk of memory), which is ~12.3MB. If the FastPFOR objects were also not pooled (ignoring the direct buffers, pretend they are pooled), it pushes the amount allocated and recycled up to ~200MB for scan of a single column.

Still, with a large enough heap it may not be a problem? As soon as I have some spare time I'll consider testing again with the value and temp arrays, leaving the codec object pool as is.

Reply:
thanks for clarifying some of the things for me. I think it is fair to measure and do whatever performs best.

I guess that not pooling temp arrays would be ok as they are short lived (specially with the optimizations done to G1GC). It might be ok to not pool value arrays as well given that we are not operating at the limits of available heap. I haven't read through the FastPFOC code yet, but sounds like it has direct buffers which are definitely better pooled.